### PR TITLE
Replace defined array_size by template version

### PR DESCRIPTION
**Pull request by khancyr**
Replace the definied array_size by the template version : 
pro : 
- type safe
- debuggable
- ensure compile time calculation
- no impact on compile time
- no impact on binary size

cons : it find to "errors" in the code : 
the first one was : 
````
../../libraries/AP_Logger/LogFile.cpp: In member function ‘void AP_Logger::Write_Current_instance(uint64_t, uint8_t, LogMessages, LogMessages)’:
../../libraries/AP_Logger/LogFile.cpp:1422:9: error: non-constant condition for static assertion
         static_assert(ARRAY_SIZE(cells.cells) == (sizeof(cell_pkt.cell_voltages) / sizeof(cell_pkt.cell_voltages[0])),
         ^~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
````
This one could be solve easily by using MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN

The second is a little more complex.
````
../../libraries/AP_ROMFS/AP_ROMFS.cpp: In static member function ‘static const uint8_t* AP_ROMFS::find_file(const char*, uint32_t&)’:
../../libraries/AP_ROMFS/AP_ROMFS.cpp:33:42: error: no matching function for call to ‘ARRAY_SIZE(const AP_ROMFS::embedded_file [0])’
     for (uint16_t i=0; i<ARRAY_SIZE(files); i++) {
                                          ^
compilation terminated due to -Wfatal-errors.
````
 On SITL with don't have ROMFS, that mean that AP_ROMFS::files is empty. Thus we cannot calculate its size. Unfortunately, we need to compile AP_ROMFS on sitl because of AP_OSD. Even if the flag WITH_SITL_OSD, is not passed, we use AP_ROMFS::find_decompress in AP_OSD_MAX7456 that is compiled by default as osd backend. My solution here was to prevent the calculation of files size and directly return the nullptr.

### CODE REVIEW

File: libraries/AP_Logger/LogFile.cpp [hunk 1]
```diff
@@ -1413,14 +1413,10 @@ void AP_Logger::Write_Current_instance(const uint64_t time_us,
             time_us             : time_us,
             voltage             : battery.voltage(battery_instance)
         };
-        for (uint8_t i = 0; i < ARRAY_SIZE(cells.cells); i++) {
+        for (uint8_t i = 0; i < MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN; i++) {
```
Comment (WickedShell) [inline libraries/AP_Logger/LogFile.cpp]: This is wrong. There are lots of reasons why the internal array may move off of the MAVLink cells to be a larger number, which is why the loop should be using the `ARRAY_SIZE`.

File: libraries/AP_Logger/LogFile.cpp [hunk 2]
```diff
@@ -1413,14 +1413,10 @@ void AP_Logger::Write_Current_instance(const uint64_t time_us,
...
             cell_pkt.cell_voltages[i] = cells.cells[i] + 1;
         }
         WriteBlock(&cell_pkt, sizeof(cell_pkt));
-
-        // check battery structure can hold all cells
-        static_assert(ARRAY_SIZE(cells.cells) == (sizeof(cell_pkt.cell_voltages) / sizeof(cell_pkt.cell_voltages[0])),
```
Comment (WickedShell) [inline libraries/AP_Logger/LogFile.cpp]: This is the key to updating the struct info. If you don't do this additional cells may not be logged silently. The assert must continue to live.

File: libraries/AP_Logger/LogFile.cpp (hunk 1 above)
Comment (khancyr) [reply libraries/AP_Logger/LogFile.cpp]: Ok I can revert that

File: libraries/AP_Logger/LogFile.cpp (hunk 2 above)
Comment (khancyr) [reply libraries/AP_Logger/LogFile.cpp]: Humm, this will be more tricky then, but I will find a soltution.

**Comment by khancyr** [comment]
Needrework until I figure out how to include the template for bootloader
what is wrong is that AP_Common should'nt be parse by waf and then in booloader build we shouldn't have the ARRAY_SIZE #define version too .

**Comment by tridge** [comment]
sorry, we've been round this before. First we had the simple macro. Then we had two systems, one macro, one simple, then we had just the simple one again.
zero length arrays are perfectly valid, and I really don't think the type benefit is worth the big increase in complexity.

**Comment by khancyr** [comment]
ok no problem

**Pull request closed without merging**